### DISCUSSION TITLE
Can you concat a relationship model.

### DISCUSSION THREAD

**Discussion by grafxflow** [rappasoft/laravel-livewire-tables#1079] [Q&A] [answered]
Can you concat a relationship model.

For example call the User model in the builder() and concat Town + City from the related Location model.

At the moment its seems I have to add Town and City and hide them in the table and concat both these rows.

**Comment by lrljoe** [comment, marked as answer]
Yep, use the setAdditionalSelects function to select the relevant fields, you can then use them in the table.

[See Manual](https://rappasoft.com/docs/laravel-livewire-tables/v2/datatable/available-methods#content-setadditionalselects)


```
public function configure(): void
{
  $this->setAdditionalSelects(['locations.townName as townName', 'locations.cityName as cityName']);
}
```

**Reply by grafxflow** [reply]
Thanks @lrljoe works like a treat.

```
Column::make('Locations ID', 'user.location.id') 
->excludeFromColumnSelect()
->hideIf(true),
    
Column::make('Town City Combined')
    ->label(
    fn($row, Column $column) => $row->townName . ', ' . $row->cityName
),
```

**Reply by lrljoe** [reply]
No problem at all 

Only thing I'd suggest is ensuring town isn't empty before displaying a comma before city.  There's some hand String Helped functions in Laravel core or you could just use a basic PHP function to display TownName, if it's not empty.

**Comment by grafxflow** [comment]
Minor question can these still be searchable and sortable?

```
return User::query()
->when($this->columnSearch['combinesTownCity'] ?? null, fn ($query, $combinesTownCity) => $query->where('locations.townName', 'like', '%' . $combinesTownCity . '%'))
->when($this->columnSearch['combinesTownCity'] ?? null, fn ($query, $combinesTownCity) => $query->where('locations.cityName', 'like', '%' . $combinesTownCity . '%'));
```

**Reply by lrljoe** [reply]
Yes to both.

**Sorting**
On the column, use the sortable method to specify how you want that sorted.
```php
    ->sortable(
        fn(Builder $query, string $direction) => $query->orderBy('townName')
    ),
```

**Searching**
```php
Column::make('Name')
    ->searchable(
        fn(Builder $query, $searchTerm) => $query->orWhere()
    ),
```

If you're going to be combining those two fields regularly by the way, you could implement a custom attribute on the model, as that's more effective for retrieval and lets you modify the display method in future in a single location.

For example, this creates an attribute on the model that returns town and city with a comma after the town if it isn't empty.
```php
    protected function townCity(): Attribute
    {
        return Attribute::make(
            get: fn (mixed $value, array $attributes) => ((! empty($attributes['town']) ? ucfirst($attributes['town']) . ',' : '') . ucfirst($attributes['city'])),
        );
    }
```

If you want to use columnSearch instead, then use orWhere to group those two together rather than two separate whens.

**Comment by grafxflow** [comment]
Great here is the amended code since using '**->label**' stopped the search functionality working so replaced it with '**->format**'.

```
public function configure(): void
{
	$this->setAdditionalSelects(['locations.townName as townName', 'locations.cityName as cityName']);
}
```

```
Column::make(__('Town City Combined'), 'location.townName')
->format(
	fn($value, $row, Column $column) => $row->townName . ', ' . $row->cityName
)
->sortable(
	fn(Builder $query, string $direction) => $query
	->orderBy('townName')
	->orderBy('cityName')
)
->searchable(
	fn(Builder $query, $searchTerm) => $query
	->orWhere('locations.townName', '=', $searchTerm)
	->orWhere('locations.cityName', '=', $searchTerm)
),
```

**Comment by lrljoe** [comment]
If you look at my attribute suggestion, you can take the (empty()) bit out of there and that'll stop spurious commas in case town is ever empty.
